assembl/graphql/vote_session.py

```diff
@@ -545,10 +623,117 @@ def mutate(root, args, context, info):
+        description_entries = args.get('description_entries')
+        discussion_id = context.matchdict['discussion_id']
+        discussion = models.Discussion.get(discussion_id)
+
+        with cls.default_db.no_autoflush as db:
+            title_ls = langstring_from_input_entries(title_entries)
+            description_ls = langstring_from_input_entries(description_entries)
+            proposal = cls(
+                discussion_id=discussion_id,
+                title=title_ls,
+                description=description_ls
+            )
+            db.add(proposal)
+            identifier = 'voteSession{}'.format(vote_session_id)
+            root_thematic = get_root_thematic_for_phase(discussion, identifier)
+            if root_thematic is None:
+                root_thematic = create_root_thematic(discussion, identifier)
+
+            order = len(root_thematic.get_children()) + 1.0
+            db.add(
+                models.IdeaLink(source=root_thematic, target=proposal,
+                                order=args.get('order', order)))
+            db.flush()
+
+        return CreateProposal(proposal=proposal)
+
+
+class UpdateProposal(graphene.Mutation):
+
+    class Input:
+        id = graphene.ID(required=True)
+        title_entries = graphene.List(LangStringEntryInput, required=True)
+        description_entries = graphene.List(LangStringEntryInput, required=True)
+        order = graphene.Float()
+
+    proposal = graphene.Field(lambda: Idea)
+
+    @staticmethod
+    @abort_transaction_on_exception
+    def mutate(root, args, context, info):
+        cls = models.Idea
+        proposal_id = args.get('id')
+        proposal_id = int(Node.from_global_id(proposal_id)[1])
+        title_entries = args.get('title_entries')
+        description_entries = args.get('description_entries')
+
+        with cls.default_db.no_autoflush as db:
+            proposal = cls.get(proposal_id)
+            require_instance_permission(CrudPermissions.UPDATE, proposal, context)
+            update_langstring_from_input_entries(
+                proposal, 'title', title_entries)
+            update_langstring_from_input_entries(
+                proposal, 'description', description_entries)
+
+            # change order if needed
+            order = args.get('order')
+            if order:
+                proposal.source_links[0].order = order
+
+            db.flush()
+
+        return UpdateProposal(proposal=proposal)
+
+
+class DeleteProposal(graphene.Mutation):
+
+    class Input:
+        id = graphene.ID(required=True)
+
+    success = graphene.Boolean()
+
+    @staticmethod
+    @abort_transaction_on_exception
+    def mutate(root, args, context, info):
+        proposal_id = args.get('id')
+        proposal_id = int(Node.from_global_id(proposal_id)[1])
+        proposal = models.Idea.get(proposal_id)
+        require_instance_permission(CrudPermissions.DELETE, proposal, context)
+        proposal.db.delete(proposal)
```

Comment:
You'd want to delete the `idea_link` (or tombstone it) as well, no?

Reply:
IdeaLink is removed when you delete the idea. We could tombstone both idea and idealink instead. Do we really want to keep proposals history?

Reply:
I tombstone now the proposal instead of deleting it in the DeleteProposal mutation.